### ISSUE TITLE
[RFC] Shardy:MPMD – an ergonomic and performant MLIR-based MPMD partitioning system

### ISSUE THREAD

**Issue by JoelWee** [openxla/shardy#586]
We've just published our RFC on Shardy:MPMD, and would love comments.

RFC: https://github.com/openxla/shardy/blob/main/rfcs/2025-06-18-mpmd-rfc.md
Read-only demo colab: https://github.com/openxla/shardy/blob/main/rfcs/2025-06-18-mpmd-rfc-read-only-colab.ipynb

Highlights from the RFC and colab:
* Shardy:MPMD is an MPMD partitioning system, aimed at enabling ergonomic and performant MPMD-partitioning while retaining single-device programming semantics.
* It comes with an experimental JAX API which compiles to a bunch of XLA programs, orchestrated by Pathways on Cloud. 
  * See the colab for an example with different pipeline schedules applied.

Let us know what you think!

Feel free to comment here, or on PR https://github.com/openxla/shardy/pull/583.

**Comment by patrick-toulme** [comment]
Very impressive. Will this be supported on XLA GPU as well?

**Comment by JoelWee** [comment]
Yes, it is supported with XLA GPU! What Shardy:MPMD does is that it creates a bunch of StableHLO programs which can then be orchestrated. I.e. so it'll work with XLA GPU / anything that would work with StableHLO :)

**Comment by MoFHeka** [comment]
Compared with 1F1B, reinforcement learning is the best and important application scenario. It's possible to deploy multiple roles in RL system easily by Shardy:MPMD?

**Comment by JoelWee** [comment]
> Any relationship between this and [torch_xla](https://github.com/pytorch/xla/issues/9019#issuecomment-2938686214)?

Not that I'm aware of – we've discussed briefly with some of the folks there about using Shardy:MPMD, but from our discussions, it seems that IFRT IR would be a better fit (unless they need the MPMD partitioning/propagation that Shardy:MPMD gives).

> Compared with 1F1B, reinforcement learning is the best and important application scenario. It's possible to deploy multiple roles in RL system easily by Shardy:MPMD?

What do you mean by RL system in this case? Do you mean e.g. the training + sampling for RL?

**Comment by MoFHeka** [comment]
@JoelWee In RL system, we usually create several Ray actors for policy model, reference model, reward model and so on. Each actor run in a training pipeline synchronously. 
Maybe I misunderstood the Shardy:MPMD, is it just for building a Local SPMD (PJRT) for pipeline parallelism?
If I want to create a multiple model training pipeline, I still need to use IFRT IR? What's the relationship between IFRT IR and Shardy:MPMD? Will IFRT IR regard Shardy:MPMD as a SPMD component? But even in single node, JAX use IFRT not PJRT.

**Comment by JoelWee** [comment]
> it seems that Shardy:MPMD need to be run in close-source Pathway

For now yes, but we're working on an open-source runtime :)

> And what's different between Pathway and McJax

The main difference is that Pathways is a single-controller runtime orchestrator, whereas McJAX is a multi-controller open-source runtime orchestrator. Aside from that, they have different performance and resilience characteristics, but that's true of any two systems.

**Comment by MoFHeka** [comment]
Brilliant! Is this on-coming runtime a single controller runtime? 
What main differences and key solutions between this runtime and the existing Ray runtime(assuming GPU objects has been supported)? 
Is it able to run SPMD with dynamic shape, I mean uncertain numbers of output tensors such as RL.
> 
> For now yes, but we're working on an open-source runtime

**Comment by G-wendy** [comment]
Hey! We are currently developing our own compiler using the shardy.mpmd dialect. However, we’ve observed that mpmd.fragment_call cannot be nested inside an mpmd.for loop. Could you help us understand the design rationale behind this restriction?

Specifically, we intend to use mpmd.for to represent microbatch—i.e. iterating over micro-batches obtained by splitting a minibatch—and within each iteration of the mpmd.for loop, invoke mpmd.fragment_call to represent different pipeline stages in pipeline parallelism.

Is this usage pattern inconsistent with the intended semantics or architectural principles of the mpmd dialect? If so, what would be the recommended way to express such a pipeline-parallel computation with micro-batching in shardy.mpmd?
